src/core/reactor.cc

```diff
@@ -1651,6 +1651,17 @@ reactor::posix_listen(socket_address sa, listen_options opts) {
 
     try {
         fd.bind(sa.u.sa, sa.length());
+
+        if (sa.is_af_unix() && opts.unix_domain_socket_permissions) {
+            // After bind the socket is created in the file system.
+            mode_t mode = static_cast<mode_t>(opts.unix_domain_socket_permissions.value());
+            auto result = ::chmod(sa.u.un.sun_path, mode);
+            if (result < 0) {
+                ::unlink(sa.u.un.sun_path);
```

Comment:
If unlink also fails for some reason (shouldn't but could), errno is overwritten.

Need to capture errno before the unlink.